examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,393 @@
+"""
+Title: Siamese Network with Triplet Loss
+Author: [hazemessamm](https://twitter.com/hazemessamm) and [Santiago L. Valdarrama](https://twitter.com/svpino)
+Date created: 2021/03/13
+Last modified: 2021/03/22
+Description: Siamese network with custom data generator and training loop.
+"""
+
+"""
+### Setup
+"""
+
+import matplotlib.pyplot as plt
+import numpy as np
+import os
+import random
+import tensorflow as tf
+
+from pathlib import Path
+from tensorflow.keras import losses, optimizers
+from tensorflow.keras import layers
+from tensorflow.keras import Model
+from tensorflow.keras import applications
+from tensorflow.keras import preprocessing
+
+target_shape = (200, 200)
+
+
+def visualize(x):
+    anchor, positive, negative = x
+    f, (ax1, ax2, ax3) = plt.subplots(1, 3)
+    ax1.imshow(anchor[0])
+    ax2.imshow(positive[0])
+    ax3.imshow(negative[0])
+
+
+"""
+# Load the dataset
+
+We are going to use the [Totally Looks Like dataset](https://drive.google.com/drive/folders/1qQJHA5m-vLMAkBfWEWgGW9n61gC_orHl). We are going to download it and unzip it inside the `~/.keras` directory.
+
+The dataset consists on two separate files:
+* `left.zip` contains the images that we will use as the anchor.
+* `right.zip` contains the images that we will use as the positive sample (an image that looks like the anchor.)
+"""
+
+cache_dir = Path(Path.home()) / ".keras"
+anchor_images_path = cache_dir / "left"
+positive_images_path = cache_dir / "right"
+
+"""shell
+gdown --id 1jvkbTr_giSP3Ru8OwGNCg6B4PvVbcO34
+gdown --id 1EzBZUb_mh_Dp_FKD0P4XiYYSd0QBH5zW
+unzip -oq left.zip -d $cache_dir
+unzip -oq right.zip -d $cache_dir
+"""
+
+"""
+Now we can load the name of every available anchor file in the unzipped directory.
+"""
+anchor_images = [str(anchor_images_path / f) for f in os.listdir(anchor_images_path)]
+
+"""
+## Preparing the data
+
+We are going to use a `tf.data` pipeline to load the data and generate the triplets
+that we need to train the Siamese network.
+"""
+
+
+def preprocess_image(filename):
+    image_string = tf.io.read_file(filename)
+    image = tf.image.decode_jpeg(image_string, channels=3)
+    image = tf.image.convert_image_dtype(image, tf.float32)
+    image = tf.image.resize(image, target_shape)
+    return image
+
+
+def generate_negative_sample(t: tf.Tensor):
+    filename = t.numpy().decode("utf-8")
+    candidate = filename
+    while candidate == filename:
+        candidate = random.choice(anchor_images).split(os.sep)[-1]
+
+    folder = random.choice([anchor_images_path, positive_images_path])
+
+    return str(folder / candidate)
+
+
+def generate_triplets(anchor_sample):
+    parts = tf.strings.split(anchor_sample, os.sep)
+    filename = parts[-1]
+
+    positive_sample = f"{str(positive_images_path)}{os.sep}" + filename
+
+    negative_sample = tf.py_function(
+        func=generate_negative_sample, inp=[filename], Tout=tf.string
```

Comment:
I debated about this for quite some time. The totally-looks-like dataset contains sequentially numbered images, so dealing with files instead of just generating the three image filenames on the fly seems a little bit overcomplicated, but I ended up doing what you suggested:

* Use `tf.data.Dataset.list_files` to load both anchor and positive images.
* Zip both together
* Generate negative image randomly without using `py_function`

The code is much simpler now. Thanks for the suggestion!